### PR TITLE
fix(segmentation): strict maxPages=0 isolation, word-safe splitting, and syntax modernization

### PR DESCRIPTION
**Pull request by ragaeeb**
### Core Bug Fixes
- **Strict Page Isolation**: Fixed a critical bug where `maxPages=0` segments could span multiple pages. The fix involves hard-clamping the breakpoint matching window to the current page boundary and enforcing position-based page attribution.
- **Overlapping Content Resolution**: Resolved an edge case where identical content at page boundaries (suffix of page N matching prefix of page N+1) confused the segmenter's transition detection.
- **Word-Safe Splitting**: Improved `maxContentLength` behavior to prefer whitespace/punctuation boundaries (`findSafeBreakPosition`) instead of chopping words in half.
- **Unicode Safety**: Integrated surrogate pair checks to prevent corruption of multi-byte characters (emojis, rare scripts) during hard splits.

### Performance & Robustness
- **O(N²) Slicing Fix**: Optimized the main breakpoint processing loop to avoid repeated slicing of massive remaining content strings.
- **O(log N) Lookups**: Robust position-to-page attribution using binary search on precomputed boundary positions.

### Major Syntactic Refactor
Full syntactic modernization of the core engine to improve readability and maintainability:
- Switched to Syntax Sugar (Optional chaining `?.`, Nullish coalescing `??`).
- Replaced manual loops with Array methods (`.some`, `.reduce`, `.map`).
- Simplified type signatures (removed redundant explicit return types).
- Applied to: `breakpoint-processor.ts`, `segmenter.ts`, `tokens.ts`, `fuzzy.ts`, `optimize-rules.ts`, and more.

### Documentation & Maintenance
- **`AGENTS.md`**: Updated with 8 new fundamental lessons for future AI collaborators.

## Summary by CodeRabbit

* **New Features**
  * Unicode-safe splitting (grapheme/surrogate aware), improved length-bounded splits, and stricter handling for one-page-per-segment (maxPages=0) with optional max-content-length support

* **Bug Fixes**
  * Stronger page-boundary detection and fixes to prevent segments from merging across pages (including overlapping-content cases)

* **Documentation**
  * Expanded debugging guidance, process examples, and security notes about regex risks

* **Tests**
  * Broader coverage for maxPages=0, overlapping pages, Unicode/surrogate edge cases, and length-bound behavior

* **Chores**
  * Package version bumped to 2.11.5

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/segmentation/replace.ts [hunk 1]
```diff
@@ -14,83 +14,51 @@ export type ReplaceRule = NonNullable<SegmentationOptions['replace']>[number];
...
+        .map((r) => ({
             pageIdSet: r.pageIds ? new Set(r.pageIds) : undefined,
-            re,
+            re: new RegExp(r.regex, normalizeReplaceFlags(r.flags)),
             replacement: r.replacement,
-        });
-    }
-    return compiled;
-};
+        }));
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply src/segmentation/replace.ts]: https://github.com/ragaeeb/flappa-doormal/pull/17/commits/c2e540538cd4c8d25474d78571a80c628fcd012c
